Add support for gradients in wxGraphicsPen

RobinD42 opened the pull request:
This PR adds support for using gradients in wxGraphicsPen.

TODO:

- [x] Add gradient related info and methods to wxGraphicsPenInfo
- [x] Cairo
- [x] OSX
- [x] GDI+
- [x] Direct2D
- [x] Documentation
- [x] Sample

See also: https://github.com/wxWidgets/Phoenix/pull/1323, https://discuss.wxpython.org/t/scalable-icons/33015

RobinD42 committed:
Add gradient-related attributes to wxGraphicsPenInfo

RobinD42 committed:
Add gradient pen support for Cairo

RobinD42 committed:
Move the new methods to the wxCairoPenBrushBaseData section of the file

vadz commented:
Meta note: GitHub now has draft PRs, that could be used instead of "WIP" label.

vadz commented on include/wx/graphics.h:
```diff
@@ -449,6 +431,95 @@ class WXDLLIMPEXP_CORE wxGraphicsGradientStops
+// wxGraphicsPenInfo describes a wxGraphicsPen
+// ----------------------------------------------------------------------------
+
+class wxGraphicsPenInfo : public wxPenInfoBase<wxGraphicsPenInfo>
```
Not sure if this really had to be moved? It looks more logical to me to have it before `wxGraphicsPen`, even though there is no actual dependency of the latter on the former.

RobinD42 replied:
That was probably due to a first pass I did for this change that I eventually mostly undid. There was something where the forward declare wasn't sufficient, but I don't see what it might have been now so I'll move it back.

RobinD42 replied:
Oh, I remember now. Adding the `wxGraphicsGradientStops` to `wxGraphicsPenInfo` is what caused the ordering conflict. I'll switch this back and then readjust for that dependency.

vadz commented on include/wx/graphics.h:
```diff
@@ -449,6 +431,95 @@ class WXDLLIMPEXP_CORE wxGraphicsGradientStops
+                   const wxColour& c1, const wxColour& c2)
+        { 
+            m_gradientType = wxGRADIENT_LINEAR;
+            m_x1 = x1; m_y1 = y1; m_x2 = x2; m_y2 = y2;
```
```suggestion
            m_x1 = x1;
            m_y1 = y1;
            m_x2 = x2;
            m_y2 = y2;
```
(here and below)

vadz commented on include/wx/graphics.h:
```diff
@@ -449,6 +431,95 @@ class WXDLLIMPEXP_CORE wxGraphicsGradientStops
+    wxGraphicsPenInfo& 
+    LinearGradient(wxDouble x1, wxDouble y1, wxDouble x2, wxDouble y2,
+                   const wxColour& c1, const wxColour& c2)
+        { 
```
Another nitpick: this block should be deindented.

vadz commented on src/generic/graphicc.cpp:
```diff
@@ -721,6 +720,52 @@ void wxCairoPenBrushBaseData::Apply( wxGraphicsContext* context )
+        );
+    }
+
+    wxASSERT_MSG(cairo_pattern_status(m_pattern) == CAIRO_STATUS_SUCCESS,
```
I see that this isn't new, but I don't think `wxASSERT_MSG` should be used here, the error should be reported in some other way (e.g. by changing the return type to `bool` and returning false).

vadz commented on src/generic/graphicc.cpp:
```diff
@@ -867,6 +913,20 @@ wxCairoPenData::wxCairoPenData( wxGraphicsRenderer* renderer, const wxGraphicsPe
         break;
     }
+
+    if (info.GetGradientType() == wxGRADIENT_LINEAR)
```
Let's use `switch ( info.GetGradientType() )`, like this (some) compiler(s) will warn us if we add new enum elements, but forget to update the switch.

vadz commented on include/wx/graphics.h:
```diff
@@ -449,6 +431,95 @@ class WXDLLIMPEXP_CORE wxGraphicsGradientStops
+    wxDouble GetY1() const { return m_y1; }
+    wxDouble GetX2() const { return m_x2; }
+    wxDouble GetY2() const { return m_y2; }
+    wxDouble GetXO() const { return m_x1; }
```
I'm not really happy about using `xo` and `yc` for parameter names, but method names should really have more intelligible names IMO.

vadz replied:
I'm certainly fine with changing the parameter names, they are not part of the API in C++, so doing this can't break anything. Thanks for cleaning this up!

FWIW I think you should be able to handle the old names as synonyms for the new ones in Python, but you certainly know more about this than I do anyhow.

vadz commented on src/generic/graphicc.cpp:
```diff
@@ -260,6 +260,13 @@ class wxCairoPenBrushBaseData : public wxGraphicsObjectRefData
 
     virtual void Apply( wxGraphicsContext* context );
 
+    void CreateLinearGradientBrush(wxDouble x1, wxDouble y1,
```
AFAICS this function (and the one below) is used for pens too, so the name is now misleading. What about `CreateXXXGradientPattern()` instead?

RobinD42 replied:
Sounds good. I was also thinking that a good idea to change those names

vadz reviewed (commented):
Thanks for the patch! I like using `wxGraphicsPenInfo`, but if I I were doing it, I'd probably go further and also use `wxGradientInfo`, that could be initialized with either `Linear()` or `Radial()` methods, i.e.

    auto penInfo = wxGraphicsPenInfo().Gradient().Linear(0, 0, 0, 100).Stops(*wxBLUE, *wxGREEN);

or something like this. I won't insist on doing it like this, but I think it could be even nicer.

Otherwise no real comments, except for the previously mentioned style nitpicks.

And it would be great to flesh out the documentation, of course, once the API is fixed.

Thanks again!

RobinD42 commented on include/wx/graphics.h:
```diff
@@ -71,6 +71,13 @@ enum wxCompositionMode
     wxCOMPOSITION_ADD /* R = S + D */
 };
 
+enum wxGradientType {
```
Opps. Agreed.

RobinD42 committed:
Update include/wx/graphics.h

fix opening brace style

Co-Authored-By: VZ <vadim@wxwidgets.org>

RobinD42 committed:
Update include/wx/graphics.h

fix style

Co-Authored-By: VZ <vadim@wxwidgets.org>

RobinD42 committed:
Reorder classes a bit to resolve ordering conflicts

RobinD42 committed:
Deindent inline code blocks

RobinD42 committed:
Rename some *GradientBrush to *GradientPattern

RobinD42 committed:
Switch to a switch statement

RobinD42 committed:
Add the rest of the accessors in wxPenInfo and wxGraphicsPenInfo,
so wxPython can see them.

RobinD42 committed:
Set the wrap mode for linear gradient brushes,
so the areas beyond the end points are properly filled with the end point colors

RobinD42 committed:
Gradients can have matrix transforms too. Updates for Cairo.

RobinD42 committed:
Adding a matrix to wxGraphicsPenInfo means the classes need reordered again

RobinD42 committed:
tweak comments

RobinD42 committed:
Add cairo_pattern_set_matrix

RobinD42 committed:
Add support for gradient pens for GDI+. API also updated for gradient transforms, but that is not working yet.

RobinD42 committed:
Create gradient brushes for pens in wxD2DRenderer

RobinD42 committed:
API updates only, to allow building.

RobinD42 committed:
Just API updates to allow building to succeed

RobinD42 committed:
Add support for setting the transform on Direct2D gradient brushes

RobinD42 committed:
Don't use a wxRect to hold floating point values!!

RobinD42 committed:
Comment out some asserts I think are incorrect. Needs verification.

RobinD42 committed:
Like Direct2D inverting the matrix is needed for gradient transforms

RobinD42 commented:
I realized that for rendering SVGs we also need to support applying a transform to gradients, so I've added that to the API as well. With that addition the SVG rendering on Cairo appears to be about as perfect as it can get. 

Direct2D works well, although it appears there may be some rounding or off-by-one errors here and there.

GDI+ still has some glitches, but I'm not sure it can be made any better.

CoreGraphics on OSX is next.

RobinD42 commented:
This is an example of a gradient rendered without a transform, and with applying the transformation matrix.

![Snap004](https://user-images.githubusercontent.com/808502/62417266-70c93300-b600-11e9-8b2b-ee7f96628c9a.png)

Commit:
Add stroking a path with a gradient on OSX, and also gradient transforms

Commit:
Make section header comments a little more consistent

Commit:
Move wxMacCoreGraphicsMatrixData so it can be used in wxMacCoreGraphicsPenBrushDataBase

Commit:
Only clone the gradient's matrix when there is one

Commit:
Remove TODO

Commit:
User more easily understandable names for the radial gradient coordinate parameters in wxGraphicsPenInfo

Commit:
undo incorrect search/replace

Commit:
(blind) compilation fix for wxQtGraphicsRenderer gradient brush methods

Commit:
Use new parameter names in the existing CreateRadialGradientBrush methods too

Commit:
documentation tweaks

Commit:
Add notes about gradient pens and gradient transforms to changes.txt

Commit:
Add example of using a gradient pen to the drawing sample

RobinD42 committed:
Fix typo

vadz commented on docs/changes.txt:
```diff
@@ -150,6 +150,13 @@ All (GUI):
   in wxPGArrayEditorDialog.
 - Fix wxPropertyGrid issues with horizontal scrolling.
 - Add wxPG_DIALOG_TITLE wxPGProperty attribute.
+- Add support for creating a wxGraphicsPen with a gradient. The gradient will 
```
Just FYI, I've actually stopped updating the change log directly to avoid merge conflicts and now keep the notes about the changes in git notes, see [this post](https://groups.google.com/d/msg/wx-dev/NVSfcnBygmk/SAQYuHTWDwAJ). But this can stay here, of course.

I would, however, try to keep it brief, as our change logs tend to be pretty long (yes, maybe the time between the releases has something to do with it), so I'd just leave
```
- Add support for creating a wxGraphicsPen with a gradient.
- Add support for applying a transformation matrix to a gradient.
```

RobinD42 replied:
That is a good idea. I hadn't known about the git notes commands, I should take a closer look and consider that for wxPython too.

vadz commented on include/wx/graphics.h:
```diff
@@ -286,6 +269,201 @@ class WXDLLIMPEXP_CORE wxGraphicsMatrix : public wxGraphicsObject
 
+// ----------------------------------------------------------------------------
+// wxGradientStop and wxGradientStops: Specify what intermediate colors are used
+// and how they are are spread out in a gradient
```
Typo
```suggestion
// and how they are spread out in a gradient
```

vadz commented on include/wx/graphics.h:
```diff
@@ -286,6 +269,201 @@ class WXDLLIMPEXP_CORE wxGraphicsMatrix : public wxGraphicsObject
+class wxGraphicsGradientStop
+{
+public:
+    wxGraphicsGradientStop(wxColour col = wxTransparentColour,
```
Do we really need `wxTransparentColour` here or would just `wxColour()` do?

RobinD42 replied:
It wouldn't have the same meaning. `wxTransparentColour` is a valid colour, but `wxColour()` is not. Also, this is preexisting code that was just moved to fix compilation errors.

vadz commented on include/wx/graphics.h:
```diff
@@ -286,6 +269,201 @@ class WXDLLIMPEXP_CORE wxGraphicsMatrix : public wxGraphicsObject
+    wxGraphicsPenInfo& 
+    LinearGradient(wxDouble x1, wxDouble y1, wxDouble x2, wxDouble y2,
+                   const wxColour& c1, const wxColour& c2,
+                   const wxGraphicsMatrix& matrix=wxNullGraphicsMatrix)
```
Tiny style fix
```suggestion
                   const wxGraphicsMatrix& matrix = wxNullGraphicsMatrix)
```

vadz commented on include/wx/graphics.h:
```diff
@@ -286,6 +269,201 @@ class WXDLLIMPEXP_CORE wxGraphicsMatrix : public wxGraphicsObject
+    wxGraphicsPenInfo& 
+    LinearGradient(wxDouble x1, wxDouble y1, wxDouble x2, wxDouble y2,
+                   const wxGraphicsGradientStops& stops,
+                   const wxGraphicsMatrix& matrix=wxNullGraphicsMatrix)
```
```suggestion
                   const wxGraphicsMatrix& matrix = wxNullGraphicsMatrix)
```

vadz commented on include/wx/graphics.h:
```diff
@@ -373,82 +551,6 @@ class WXDLLIMPEXP_CORE wxGraphicsPath : public wxGraphicsObject
 extern WXDLLIMPEXP_DATA_CORE(wxGraphicsPath) wxNullGraphicsPath;
 
 
-// Describes a single gradient stop.
```
It's still a bit annoying that this has moved around. We can use `git diff --color-moved`, of course, but e.g. GitHub doesn't detect moves, so if it were possible to move it back, it would result in a better diff here.

vadz commented on interface/wx/graphics.h:
```diff
@@ -305,6 +305,13 @@ enum wxCompositionMode
     wxCOMPOSITION_ADD  /**< @e R = @e S + @e D */
 };
 
+enum wxGradientType {
```
Should add doc comment before it if only to put `@since 3.1.3` into it.

vadz commented on interface/wx/graphics.h:
```diff
@@ -633,46 +640,54 @@ class wxGraphicsContext : public wxGraphicsObject
 
         The version taking wxGraphicsGradientStops is new in wxWidgets 2.9.1.
+
+        The ability to apply a transformation matrix to the gradient was added in 3.1.3
```
I think we mostly use wording like `@a matrix parameter is only available since wxWidgets 3.1.3`. I'm not really sure, but I'd just like to make it consistent with the existing documentation.

vadz commented on src/generic/graphicc.cpp:
```diff
@@ -721,6 +722,66 @@ void wxCairoPenBrushBaseData::Apply( wxGraphicsContext* context )
+{
+    m_pattern = cairo_pattern_create_linear(x1,y1,x2,y2);
+
+    if (! matrix.IsNull())
```
Another style nitpick:
```suggestion
    if ( !matrix.IsNull() )
```

vadz commented on src/generic/graphicc.cpp:
```diff
@@ -721,6 +722,66 @@ void wxCairoPenBrushBaseData::Apply( wxGraphicsContext* context )
+{
+    m_pattern = cairo_pattern_create_radial(startX,startY,0.0,endX,endY,radius);
+
+    if (! matrix.IsNull())
```
```suggestion
    if ( !matrix.IsNull() )
```

vadz commented on src/generic/graphicc.cpp:
```diff
@@ -721,6 +722,66 @@ void wxCairoPenBrushBaseData::Apply( wxGraphicsContext* context )
+
+    if (! matrix.IsNull())
+    {
+        cairo_matrix_t m = *((cairo_matrix_t*) matrix.GetNativeMatrix());
```
Ugh, this is really ugly. We should add some inline helper returning `cairo_matrix_t` (but this comment is not really aimed at this PR, it can be done later).

vadz commented on src/msw/graphics.cpp:
```diff
@@ -255,55 +255,66 @@ public :
+{
+public:
+    wxGDIPlusPenBrushBaseData(wxGraphicsRenderer* renderer);
+    ~wxGDIPlusPenBrushBaseData();
```
Doesn't really change anything, but, for consistency:
```suggestion
    virtual ~wxGDIPlusPenBrushBaseData();
```

More importantly, I think it should be protected and not public as objects of this class are not supposed to be deleted, only `DecRef()`'d until they die on their own.

vadz commented on src/msw/graphics.cpp:
```diff
@@ -255,55 +255,66 @@ public :
     ~wxGDIPlusPenData();
 
-    void Init();
+    virtual void Init();
```
```suggestion
    virtual void Init() wxOVERRIDE;
```

vadz commented on src/msw/graphics.cpp:
```diff
@@ -255,55 +255,66 @@ public :
+    void CreateLinearGradientBrush(wxDouble x1, wxDouble y1,
+                                   wxDouble x2, wxDouble y2,
+                                   const wxGraphicsGradientStops& stops,
+                                   const wxGraphicsMatrix& matrix=wxNullGraphicsMatrix);
```
```suggestion
                                   const wxGraphicsMatrix& matrix = wxNullGraphicsMatrix);
```

vadz commented on src/msw/graphics.cpp:
```diff
@@ -255,55 +255,66 @@ public :
+                                   wxDouble endX, wxDouble endY,
+                                   wxDouble radius,
+                                   const wxGraphicsGradientStops& stops,
+                                   const wxGraphicsMatrix& matrix=wxNullGraphicsMatrix);
```
```suggestion
                                   const wxGraphicsMatrix& matrix = wxNullGraphicsMatrix);
```

vadz commented on src/msw/graphics.cpp:
```diff
@@ -669,27 +682,153 @@ private :
+    brush->SetWrapMode(WrapModeTileFlipXY);
+
+    // Apply the matrix if there is one
+    if (! matrix.IsNull())
```
```suggestion
    if ( !matrix.IsNull() )
```

vadz commented on src/msw/graphics.cpp:
```diff
@@ -669,27 +682,153 @@ private :
+    // the area outside of the gradient's radius is not getting painted. 
+
+    // Apply the matrix if there is one
+    if (! matrix.IsNull())
```
```suggestion
    if ( !matrix.IsNull() )
```

vadz commented on src/msw/graphicsd2d.cpp:
```diff
@@ -1772,9 +1772,12 @@ void wxD2DPathData::AddPath(const wxGraphicsPathData* path)
     // Raise warning if appended path has an open non-empty sub-path.
-    wxASSERT_MSG( pathSrc->IsStateSafeForFlush(),
-        wxS("Sub-path in appended path should be closed prior to this operation") );
+    // wxASSERT_MSG( pathSrc->IsStateSafeForFlush(),
```
Let's remove it if it's really not needed, leaving it commented out is not really useful. OTOH I'm not completely sure it's not needed... What is the example of a path for which it is wrongly triggered?

@a-wi could you please look at this?

RobinD42 replied:
Here is an example:

![Snap001](https://user-images.githubusercontent.com/808502/64360908-4cc28e00-cfc0-11e9-85e9-42efc790dfb5.png)

If this path is closed then a line will be drawn from the trailing end to the beginning of the path, which is not correct in this case.

RobinD42 replied:
Here is the SVG file for the above image if anybody is curious. GitHub won't let me upload the file, but it's short so I'll just paste it.

```
<?xml version="1.0" encoding="utf-8"?>
<svg viewBox="0 0 200 200" xmlns="http://www.w3.org/2000/svg">
  <polyline style="fill: none; stroke-width: 3px; stroke: rgb(0, 21, 255); stroke-linecap: round; stroke-linejoin: round;" points="26.576 22.775 177.684 22.775 177.684 58.695 25.673 58.695 25.673 88.195 173.737 88.195 173.737 129.814 25.764 129.814 25.764 169.099 175.947 169.099"/>
</svg>
```

RobinD42 replied:
I'm going to go ahead and remove the commented out asserts. Let's discuss on wx-dev if it's not correct.

vadz reviewed (commented):
Thanks a lot, this is certainly a nice addition to the graphics API!

Public API looks perfectly fine to me and I don't have any real comments about Cairo/MSW changes. I didn't really look at CG ones because I don't know anything about CG anyhow, so the fact that it works is good enough for me.

Except for the multitude of micro style fixes, the only real question is about the asserts in D2D code.

RobinD42 committed:
Various typo and coding style fixes

RobinD42 committed:
Another typo fix

RobinD42 committed:
Add another missed wxOVERRIDE

RobinD42 committed:
Remove misplaced wxOVERRIDE

RobinD42 committed:
Remove commented-out asserts
